### PR TITLE
validate client secret expired or not

### PR DESCRIPTION
**Pull request by doctormacky**
validate the client secret is expired or not.
see the details [here](https://github.com/spring-projects/spring-authorization-server/issues/850)

### CODE REVIEW

**Comment by doctormacky** [comment]
@jgrandja Please help to review this PR. thanks!

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/ClientSecretAuthenticationProvider.java [hunk 1]
```diff
@@ -93,6 +95,11 @@ public Authentication authenticate(Authentication authentication) throws Authent
 			throwInvalidClient(OAuth2ParameterNames.CLIENT_ID);
 		}
 
+		Instant expiredAt = registeredClient.getClientSecretExpiresAt();
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/ClientSecretAuthenticationProvider.java:98]: Please move this validation below `if (!this.passwordEncoder.matches(clientSecret, registeredClient.getClientSecret()))`

Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/ClientSecretAuthenticationProvider.java:98]: Remove the variable `expiredAt` and just use `registeredClient.getClientSecretExpiresAt()`

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/ClientSecretAuthenticationProvider.java [hunk 2]
```diff
@@ -93,6 +95,11 @@ public Authentication authenticate(Authentication authentication) throws Authent
...
+		if (expiredAt!=null && Instant.now().isAfter(expiredAt)) {
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/ClientSecretAuthenticationProvider.java:99]: Add spaces between the comparison `expiredAt!=null` so it's consistent with other code

**Review by jgrandja — CHANGES_REQUESTED** [review]
Thanks for the PR @doctormacky. Please see review comments.

Also, please add a test in `ClientSecretAuthenticationProviderTests`.

**Comment by doctormacky** [comment]
@jgrandja Please help to review it again.  thanks.

**Comment by doctormacky** [comment]
@jgrandja Any progress on this PR ?

File: oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/client/TestRegisteredClients.java [hunk 3]
```diff
@@ -64,4 +64,20 @@ public static RegisteredClient.Builder registeredPublicClient() {
 				.scope("scope1")
 				.clientSettings(ClientSettings.builder().requireProofKey(true).build());
 	}
+	
+	public static RegisteredClient.Builder registeredClient4() {
```
Comment (jgrandja) [inline oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/client/TestRegisteredClients.java:68]: Please remove this and reuse `TestRegisteredClients.registeredClient()`, which returns the `Builder` and allows you to update `clientSecretExpiresAt()`

**Review by jgrandja — CHANGES_REQUESTED** [review]
Thanks for the updates @doctormacky.

One minor change left and we'll be ready to merge.

Can you also rebase off latest `0.4.x` and squash the commits.
Thanks!

**Comment by jgrandja** [comment]
Thanks for the updates @doctormacky. This is now merged.

**Comment by doctormacky** [comment]
> Thanks for the updates @doctormacky. This is now merged.

Thanks @jgrandja  It's my pleasure.

**Pull request closed without merging**